### ISSUE TITLE
Request commit model for cifar10

### ISSUE THREAD

**Issue by jack-kjlee** [ruizhoud/DistributionLoss#3]
Can you commit model for cifar10 described in the paper xC-xC-MP-2xC-2xC-MP-4xC-4xC-10C-GP?
I have some difficulties to reproduce result of the papers.

**Comment by jack-kjlee** [comment]
Below code is my model.
First layer is also binarized layer since I use binarized encoded cifar10 data.
With original BNN code, my model achieved 85.5% accuracy.
However, accuracy was dropped to 74~75% when I applied distribution loss.
I applied Distribution loss in all convolution layers, not fc layer

**[DL]**
k1 = 0.25
k2 = 1
distrloss1 = (torch.min(1 - mean - (k1*std), 1 + mean - (k1*std)).clamp(min=0) ** 2).mean() + (((k1*std) - 1).clamp(min=0) ** 2).mean()
distrloss2 = ((mean.abs() - (k2*std)).clamp(min=0) ** 2).mean()

**[Configuration]**
Total epoch = 200
DL epoch = 60
DL lambda = 2
Optimizer = Adam
Learning Rate: start from 0.01 (described in below model code)

**[Model code]**
`
def __init__(self, tb_prefix, writer):
    super(MYModel, self).__init__()
       
    self.activation_func = HardTanh_bin
    self.channels = [128, 128, 128, 128, 128, 128, 128]
    
    self.features0 = nn.Sequential(
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[0], self.channels[1], kernel_size=3, stride=1, padding=0, bias=False, groups=16),
        nn.BatchNorm2d(self.channels[1]), #, momentum=None
    )
    self.features1 = nn.Sequential(
        self.activation_func(),
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[1], self.channels[2], kernel_size=3, stride=1, padding=0, bias=False, groups=16),
        nn.BatchNorm2d(self.channels[2]),
        nn.MaxPool2d(kernel_size=2, stride=2),
    )
    
    self.features2 = nn.Sequential(
        self.activation_func(),
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[2], self.channels[3], kernel_size=3, stride=1, padding=0, bias=False, groups=4),
        nn.BatchNorm2d(self.channels[3]),
    )
    self.features3 = nn.Sequential(
        self.activation_func(),
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[3], self.channels[4], kernel_size=3, stride=1, padding=0, bias=False, groups=4),
        nn.BatchNorm2d(self.channels[4]),
        nn.MaxPool2d(kernel_size=2, stride=2),
    )
    
    self.features4 = nn.Sequential(
        self.activation_func(),
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[4], self.channels[5], kernel_size=3, stride=1, padding=0, bias=False),
        nn.BatchNorm2d(self.channels[5]),
    )
    self.features5 = nn.Sequential(
        self.activation_func(),
        nn.ReplicationPad2d(1),
        BinarizeConv2d(self.channels[5], self.channels[6], kernel_size=3, stride=1, padding=0, bias=False),
        nn.BatchNorm2d(self.channels[6]),
        nn.MaxPool2d(kernel_size=2, stride=2),
    )
    

    self.neurons = [self.channels[6] * 4 * 4, 10]
    self.classifier0 = nn.Sequential(
        self.activation_func(),
        BinarizeLinear(self.neurons[0], self.neurons[1], bias=True),
        nn.BatchNorm1d(self.neurons[1], affine=False), #, momentum=None
        nn.LogSoftmax()
    )
    
    # Wight Initialization
    for m in self.modules():
        if isinstance(m, BinarizeConv2d):
            # Xavier Initialization
            nn.init.xavier_normal_(m.weight.data)
            
        if isinstance(m, BinarizeLinear):                
            # Xavier Initialization
            nn.init.xavier_normal_(m.weight.data)
            m.bias.data.fill_(0)

    self.distrloss_layers = []
    for i in range(0, 6):
        self.distrloss_layers.append(Distrloss_layer(self.channels[i]))
        
    for i in range(0, 1):
        self.distrloss_layers.append(Distrloss_layer(self.neurons[1]))

    self.regime = {
        0: {'optimizer': 'Adam', 'lr': 1e-2},
        20: {'lr': 5e-3},
        40: {'lr': 1e-3},
        60: {'lr': 5e-4},
        80: {'lr': 1e-4},
        100: {'lr': 5e-5},
        120: {'lr': 1e-5},
        140: {'lr': 5e-6},
    }
    self.input_transform = {
        'train': transforms.Compose([
            transforms.RandomCrop(32, padding=4),
            transforms.RandomHorizontalFlip(),
            transforms.ToTensor(),
        ]),
        'eval': transforms.Compose([
            transforms.ToTensor(),
        ])
    }

def forward(self, x):
    loss = []
    
    x = self.features0(x)
    loss.append(self.distrloss_layers[0](x))
    
    x = self.features1(x)
    loss.append(self.distrloss_layers[1](x))
    
    x = self.features2(x)
    loss.append(self.distrloss_layers[2](x))
    
    x = self.features3(x)
    loss.append(self.distrloss_layers[3](x))
    
    x = self.features4(x)
    loss.append(self.distrloss_layers[4](x))
    
    x = self.features5(x)
    loss.append(self.distrloss_layers[5](x))

    x = x.view(-1, self.channels[6] * 4 * 4)

    x = self.classifier0(x)

    distrloss1 = sum([ele[0] for ele in loss]) / len(loss)
    distrloss2 = sum([ele[1] for ele in loss]) / len(loss)
    
    self.n += 1
    
    return x, distrloss1.view(1,1), distrloss2.view(1,1)
`

Could you give me any advise?
Thanks.

**Comment by jack-kjlee** [comment]
I'm really waiting for an answer.
It is hard for me to find tconfiguration of distribution loss of cifar10 model (Network ID1: 1~5 in Table3).

**Comment by ruizhoud** [comment]
thanks for reminding.. can you hold on for one week? I will try to commit it next week..

**Comment by jack-kjlee** [comment]
> thanks for reminding.. can you hold on for one week? I will try to commit it next week..

Thanks for answer 👍

**Comment by jack-kjlee** [comment]
Sorry for my reply again. Could you tell me progress? I need to re-schedule my plan..

**Comment by liyue2ppy** [comment]
@jack-kjlee Have you solve the problem? I try the distribution loss in BNNs of a resnet18 architectue on CIFAR100, but get a 4% drop in accuracy. Do you know the proper configuration?
